src/bufr/DataObject.h

```diff
@@ -372,6 +399,131 @@ namespace Ingester
+        {
+            data_ = std::vector<T>(data.begin(), data.end());
+            std::replace(data_.begin(),
+                         data_.end(),
+                         static_cast<T>(dataMissingValue),
+                         missingValue());
+        }
+
+        /// \brief Set the data associated with this data object (string DataObject).
+        /// \param data - double vector of raw data
+        /// \param dataMissingValue - The number that represents missing values within the raw data
+        template<typename U = void>
+        void _setData(
+            const std::vector<double>& data,
+            double dataMissingValue,
+            typename std::enable_if<std::is_same<T, std::string>::value, U>::type* = nullptr)
+        {
+            data_ = std::vector<std::string>();
+            auto charPtr = reinterpret_cast<const char*>(data.data());
+            for (size_t row_idx = 0; row_idx < data.size(); row_idx++)
+            {
+                if (data[row_idx] != dataMissingValue)
+                {
+                    std::string str = std::string(
+                        charPtr + row_idx * sizeof(double), sizeof(double));
+
+                    // trim trailing whitespace from str
+                    str.erase(std::find_if(str.rbegin(), str.rend(),
+                                           [](char c){ return !std::isspace(c); }).base(),
+                              str.end());
+
+                    data_.push_back(str);
+                }
+                else
+                {
+                    data_.push_back("");
+                }
+            }
+        }
+
+        /// \brief Multiply the stored values in this data object by a scalar.
+        /// \param val Scalar to multiply to the data..
+        void multiplyBy(double val) final
+        {
+            _multiplyBy(val);
+        }
+
+        /// \brief Multiply the stored values in this data object by a scalar (numeric version).
+        /// \param val Scalar to multiply to the data.
+        template<typename U = void>
+        void _multiplyBy(double val,
+                         typename std::enable_if<std::is_arithmetic<T>::value, U>::type* = nullptr)
+        {
+            if (typeid(T) == typeid(float) ||   // NOLINT
+                typeid(T) == typeid(double) ||  // NOLINT
+                trunc(val) == val)
+            {
+                for (size_t i = 0; i < data_.size(); i++)
+                {
+                    if (data_[i] != missingValue())
+                    {
+                        data_[i] = static_cast<T>(static_cast<double>(data_[i]) * val);
+                    }
+                }
+            }
+            else
+            {
+                std::ostringstream str;
+                str << "Multiplying integer field \"" << fieldName_ << "\" with a non-integer is ";
+                str << "illegal. Please convert it to a float or double.";
+                throw std::runtime_error(str.str());
```

Comment:
I realize that I do not fully understand the problems or context you are trying to solve with the data typing improvements in this PR. I apologize for having this limited understanding, but to me these new functions you added (setData, multiplyBy, offsetBy, etc.) seem overly complex since they seem to be trying to replace the compile time type matching/checking you get with templates with runtime type checking.

For example the multiplyBy routine here is set to use a factor that is of type double. Why not use a factor of type T instead. This removes the need to do the checking and type conversions (static_casting). Ie, the compiler will create a version with float, another with int, etc. according to what the rest of the code instantiates. In this scheme the non-numeric case (string) could be handled with a template specialization of DataObject that simply throws exceptions if you try to call multiplyBy with strings.